index.js

```diff
@@ -154,36 +154,37 @@ function findTag(gitPath, sha) {
 
 var LAST_TAG_CACHE = {};
 
-function findLastTagCached(gitPath, sha, commitsSinceLastTag) {
+function findLastTagCached(gitPath, sha) {
   if(!LAST_TAG_CACHE[gitPath]) {
     LAST_TAG_CACHE[gitPath] = {};
   }
 
   if(!LAST_TAG_CACHE[gitPath][sha]) {
-    LAST_TAG_CACHE[gitPath][sha] = findLastTag(gitPath, sha, commitsSinceLastTag);
+    LAST_TAG_CACHE[gitPath][sha] = findLastTag(gitPath, sha);
   }
 
   return LAST_TAG_CACHE[gitPath][sha];
 }
 
-function findLastTag(gitPath, sha, commitsSinceLastTag) {
-  commitsSinceLastTag = commitsSinceLastTag || 0;
-  var tag = findTag(gitPath, sha);
-  if(!tag) {
+function findLastTag(gitPath, sha) {
+  var queue = [{ sha: sha, depth: 0 }];
+  while (queue.length) {
+    var element = queue.shift();
+    var tag = findTag(gitPath, element.sha);
+    if (tag) {
+      return {
+        tag: tag,
+        commitsSinceLastTag: element.depth
+      };
+    }
     var commitData = getCommitData(gitPath, sha);
-    if(!commitData) {
-      return { tag: null, commitsSinceLastTag: Infinity };
+    if (commitData && commitData.parents) {
+      for (var i = 0; i < commitData.parents.length; i++) {
+        queue.push({ sha: commitData.parents[i], depth: element.depth + 1 });
+      }
     }
```

Comment:
Looks like we don't keep track of SHAs we've seen.  I think we'll have a degenerate case where we do repeat searches once we get to the common ancestor.  eg

```
A1   B1
|    |
A2   B2
|    |
A3   B3
 \ /
  C1
  |
  C2
  |
  C3 tag: ohai
```

will search C1 and C2 twice.

I don't think this actually matters very much since it would only be costly in a degenerate case, but it might be nice to leave a comment to explain the potential optimization

Reply:
This is a good point and I think it doesn't happen only in degenerate cases. I believe this will happen for every merge commit. Keeping track of SHAs probably wouldn't cause much overhead and would definitely improve the performance in many cases. I'll test time when I have some time.